src/Functions/array/arrayElement.cpp

```diff
@@ -819,11 +818,40 @@ void FunctionArrayElement::executeMatchKeyToIndex(
     const Offsets & offsets, PaddedPODArray<UInt64> & matched_idxs, const Matcher & matcher)
 {
     size_t rows = offsets.size();
-    for (size_t i = 0; i < rows; ++i)
+    size_t matched_pos = 0;
+    bool matched = false;
+    for (size_t j = 0, end = offsets[0]; j < end; ++j)
     {
-        bool matched = false;
-        size_t begin = offsets[i - 1];
-        size_t end = offsets[i];
+        if (matcher.match(j, 0))
+        {
+            matched_idxs.push_back(j + 1);
+            matched = true;
+            matched_pos = end + j;
+            break;
+        }
+    }
+    if (!matched)
+        matched_idxs.push_back(0);
+    /// In pratice, map keys are usually in the same order, it is worth a trying to
+    /// predicate the next key position as previous. So it can save a lot of comparisions.
+    size_t i = 1;
+    for (; i < rows; ++i)
+    {
+        const auto & begin = offsets[i - 1];
+        const auto & end = offsets[i];
+        if (matched_pos < end && matcher.match(matched_pos, i))
```

Comment:
Here `matched_pos < end` is a bit confusing. I think the logic is correct, but maybe it worth to consider renaming `matched_pos` to `expected_match_pos`.

Reply:
done